lib/pharos/phases/configure_master.rb

```diff
@@ -118,36 +134,108 @@ def generate_config
         end
 
         config['apiServerExtraArgs'] = {}
         config['apiServerExtraVolumes'] = []
 
         # Only if authentication token webhook option are given
-        if @config.authentication&.token_webhook
-          config['apiServerExtraArgs'].merge!(authentication_token_webhook_args(@config.authentication.token_webhook.cache_ttl))
-          config['apiServerExtraVolumes'] += volume_mounts_for_authentication_token_webhook
-        end
+        configure_token_webhook(config) if @config.authentication&.token_webhook
 
         # Configure audit related things if needed
-        if @config.audit&.server
-          config['apiServerExtraArgs'].merge!(
-            "audit-webhook-config-file" => AUDIT_CFG_DIR + '/webhook.yml',
-            "audit-policy-file" => AUDIT_CFG_DIR + '/policy.yml'
-          )
-          config['apiServerExtraVolumes'] += volume_mounts_for_audit_webhook
-        end
+        configure_audit_webhook(config) if @config.audit&.server
 
         config
       end
 
+      # @return [Array<String>]
+      def build_extra_sans
+        extra_sans = Set.new(['127.0.0.1'])
+        @config.master_hosts.each do |h|
+          extra_sans += [h.address, h.private_address].compact.uniq
+        end
+
+        extra_sans.to_a
+      end
+
+      # Copies certificates from leading master
+      def copy_kube_certs
+        leader = @config.master_leader
+        leader_ssh = Pharos::SSH::Client.for_host(leader)
+        @ssh.exec!("sudo mkdir -p #{KUBE_DIR}/pki")
+        %w(ca.crt ca.key sa.key sa.pub).each do |file|
+          path = File.join(KUBE_DIR, 'pki', file)
+          contents = leader_ssh.file(path).read
+          @ssh.file(path).write(contents)
+          @ssh.exec!("sudo chmod 0400 #{path}")
+        end
+      end
+
+      # @param config [Pharos::Config]
+      def configure_internal_etcd(config)
+        endpoints = @config.etcd_hosts.map { |h|
+          "https://#{h.peer_address}:2379"
+        }
+        config['etcd'] = {
+          'endpoints' => endpoints
+        }
+
+        config['etcd']['certFile'] = '/etc/pharos/pki/etcd/client.pem'
+        config['etcd']['caFile'] = '/etc/pharos/pki/ca.pem'
+        config['etcd']['keyFile'] = '/etc/pharos/pki/etcd/client-key.pem'
+      end
+
+      # TODO: lock down permissions on key
+      def copy_external_etcd_certs
+        @ssh.exec!('sudo mkdir -p /etc/pharos/etcd')
+        @ssh.file('/etc/pharos/etcd/ca-certificate.pem').write(File.open(@config.etcd.ca_certificate))
+        @ssh.file('/etc/pharos/etcd/certificate.pem').write(File.open(@config.etcd.certificate))
+        @ssh.file('/etc/pharos/etcd/certificate-key.pem').write(File.read(@config.etcd.key))
+      end
+
+      def copy_internal_etcd_certs
+        @ssh.exec!('sudo mkdir -p /etc/pharos/pki/etcd')
+        return if @ssh.file('/etc/pharos/pki/etcd/client.pem').exist?
+
+        etcd_peer = @config.etcd_hosts[0]
+        etcd_ssh = Pharos::SSH::Client.for_host(etcd_peer)
```

Comment:
Hmm... isn't this now most likely going to be `localhost`, for the case of a `cluster.yaml` with a single master node?

Reply:
Unless you have separate `etcd` nodes, then the etcd certs are already going to be in place after the `ConfigureEtcd` phase?

Feels like this operation should only be necessary in some specific circumstances, so maybe it shouldn't happen unconditionally.